ledger/src/blockstore_processor.rs

```diff
@@ -415,18 +423,22 @@ fn process_batches(
 
 fn schedule_batches_for_execution(
     bank: &BankWithScheduler,
-    batches: &[TransactionBatchWithIndexes<SanitizedTransaction>],
+    locked_entries: impl Iterator<Item = LockedTransactionsWithIndexes<SanitizedTransaction>>,
 ) -> Result<()> {
-    for TransactionBatchWithIndexes {
-        batch,
-        transaction_indexes,
-    } in batches
+    for LockedTransactionsWithIndexes {
+        lock_results,
+        transactions,
+        starting_index,
+    } in locked_entries
     {
+        // unlock before sending to scheduler.
+        bank.unlock_accounts(transactions.iter().zip(lock_results.iter()));
```

Comment:
Why is it okay to unlock accounts here? Do they not need to complete execution?

Reply:
unified scheduler does its own lock management.
@ryoqun correct me if I'm wrong, but afaik the batches in original code are dropped soon after (and unlocked) since scheduling here is non-blocking anyway.

Reply:
Probably outside the scope of this PR, but I could use an education on how this locking mechanism works. Like how do we make sure we execute things in the right order on replay? Do we guarantee the batches get executed in order once scheduled?

Reply:
> Like how do we make sure we execute things in the right order on replay?

Unified scheduler realizes this with 2 things: the event loop and `SchedulingStateMachine`.

The event loop is high-level driver of the state machine, starting from receiving tasks (= transactions) via `new_task_receiver`, which is sent from this blockstore code:

https://github.com/anza-xyz/agave/blob/17346cdfb292dff1ee538281b08859d3acf7a139/unified-scheduler-pool/src/lib.rs#L1031-L1081

and this is relevant impl code of `SchedulingStateMachine`:

https://github.com/anza-xyz/agave/blob/17346cdfb292dff1ee538281b08859d3acf7a139/unified-scheduler-logic/src/lib.rs#L664-L755

> Do we guarantee the batches get executed in order once scheduled?

yes, while maximally parallelizing executing tasks. _order_ here is the order transactions are sent over the crossbeam channel.

Reply:
it uses crossbeam order, but it also verifies the `index` is never out of order for conflicts, iirc.

index order is guaranteed with the way we send process entries in order (until conflict)

Reply:
well, at the last minute, we removed the index verification: https://github.com/solana-labs/solana/pull/35286#discussion_r1515662275

> index order is guaranteed with the way we send process entries in order (until conflict)

this is true.